lib/fizzy/execute.cpp

```diff
@@ -378,6 +378,55 @@ execution_result execute(Instance& instance, FuncIdx func_idx, std::vector<uint6
             labels.push_back(label);
             break;
         }
+        case Instr::if_:
+        {
+            if (static_cast<uint32_t>(stack.pop() != 0))
+            {
+                const auto arity = read<uint8_t>(immediates);
+                const auto target_pc = read<uint32_t>(immediates);
+                const auto target_imm = read<uint32_t>(immediates);
+                read<uint32_t>(immediates);  // Omit the immediates for else instruction.
+                read<uint32_t>(immediates);  // Omit the immediates for else instruction.
+                LabelContext label{code.instructions.data() + target_pc,
+                    code.immediates.data() + target_imm, arity, stack.size()};
+                labels.emplace_back(label);
+            }
+            else
+            {
+                const auto arity = read<uint8_t>(immediates);
+                const auto target_pc = read<uint32_t>(immediates);
+                const auto target_imm = read<uint32_t>(immediates);
+                const auto target_else_pc = read<uint32_t>(immediates);
+                const auto target_else_imm = read<uint32_t>(immediates);
+
+                if (target_else_pc != 0 && target_else_imm != 0)  // If else block defined.
+                {
+                    LabelContext label{code.instructions.data() + target_pc,
+                        code.immediates.data() + target_imm, arity, stack.size()};
+                    labels.emplace_back(label);
+                    pc = code.instructions.data() + target_else_pc;
+                    immediates = code.immediates.data() + target_else_imm;
+                }
+                else  // If else block not defined go to end of if.
+                {
+                    assert(arity == 0);  // if without else cannot have type signature.
```

Comment:
Could you point where is this in the spec?

Reply:
WebAssembly Studio returns the error when parsing `if` instructions. Unfortunately I cannot find any information in spec. I can remove the assert if you decide so.

Reply:
In simplistic view, both `if` and `else` blocks must have the same type. If `else` is omitted the type is `[]` therefore `if`'s type must be `[]`.
https://webassembly.github.io/spec/core/valid/instructions.html#valid-if